File: tools/cockpit/backend/freshness.py
"""Staleness classification per source cadence (cockpit hand-paired copy).

CANONICAL SOURCE: deploy/energy-stack/hvac-scheduler/freshness.py.
This file is byte-identical to the canonical except for this header
docstring. Drift is enforced by the CI workflow at
.github/workflows/check-freshness-drift.yml — any edit must land in
BOTH files in the same PR or CI will fail.

Per spec §3.1: `comed.prices` fresh threshold of 7 min is the
controller's downgrade-recency cutoff. Cockpit mirrors so the operator
sees exactly the same actionability the controller does.

`hvac.actions` is event-driven and NOT a staleness signal — Action
node uses NOT-FIRED-THIS-TICK / last-fire / APPLIED / SHADOW semantics.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

Freshness = Literal["fresh", "warn", "stale", "missing"]


@dataclass(frozen=True)
class Thresholds:
    fresh_max_ms: int
    warn_max_ms: int
    stale_max_ms: int


def _min(n: float) -> int:
    return int(n * 60 * 1000)


def _hr(n: float) -> int:
    return int(n * 60 * 60 * 1000)


THRESHOLDS: dict[str, Thresholds] = {
    "decision_trace.price_overlay_eval": Thresholds(_min(6), _min(10), _min(15)),
    "decision_trace.layer_resolution":   Thresholds(_min(6), _min(10), _min(15)),
    "decision_trace.day_type_decision":  Thresholds(_hr(16), _hr(30), _hr(72)),
    "decision_trace.precool_decision":   Thresholds(_hr(26), _hr(40), _hr(72)),
    "hvac.arm_mode":                     Thresholds(_min(6), _min(10), _min(15)),
    "hvac.thermostat":                   Thresholds(_min(12), _min(20), _min(30)),
    # 7-min fresh = controller's downgrade-recency cutoff. Cockpit mirrors
    # so operator sees exactly the same actionability the controller does.
    # Bucket-age sawtooth typically spans 6-11 min between publishes, so
    # the cockpit's 'fresh' indicator naturally cycles green→warn→green
    # every 5-min publish cycle. Warn does NOT indicate a feed problem —
    # it indicates the controller would refuse a downgrade decision if
    # asked this tick. See spec §3.1 + §6.
    "comed.prices":                      Thresholds(_min(7), _min(16), _min(30)),
    "nws.forecast":                      Thresholds(_min(35), _min(90), _hr(12)),
    "pjm.load_forecast":                 Thresholds(_hr(14), _hr(28), _hr(50)),
    "pjm.rt_hrl_lmps":                   Thresholds(_min(75), _hr(3), _hr(12)),
    "refoss.channel":                    Thresholds(_min(1), _min(3), _min(10)),
    "eagle.meter":                       Thresholds(_min(1), _min(3), _min(10)),
}


def classify(source: str, age_ms: int) -> Freshness:
    """Map an age in ms to a freshness bucket per source cadence."""
    t = THRESHOLDS.get(source)
    if t is None:
        return "fresh"
    if age_ms <= t.fresh_max_ms:
        return "fresh"
    if age_ms <= t.warn_max_ms:
        return "warn"
    if age_ms <= t.stale_max_ms:
        return "stale"
    return "missing"

File: tools/cockpit/frontend/src/freshness.ts
// Staleness thresholds per source cadence. Used by future Phase 3 backend
// AND by any frontend code that derives `freshness` from a raw timestamp.
//
// IMPORTANT: hvac.actions is event-driven and NOT a staleness signal.
// Action node renders NOT-FIRED-THIS-TICK / last-fire / APPLIED / SHADOW
// semantics. Liveness comes from hvac.arm_mode + hvac.heartbeat + 5-min
// trace cadence.

import type { Freshness } from './types'

export interface FreshnessThresholds {
  fresh_max_ms: number
  warn_max_ms: number
  stale_max_ms: number
}

const min = (n: number) => n * 60 * 1000
const hr = (n: number) => n * 60 * 60 * 1000

export const FRESHNESS_THRESHOLDS: Record<string, FreshnessThresholds> = {
  'decision_trace.price_overlay_eval': {
    fresh_max_ms: min(6),
    warn_max_ms: min(10),
    stale_max_ms: min(15),
  },
  'decision_trace.layer_resolution': {
    fresh_max_ms: min(6),
    warn_max_ms: min(10),
    stale_max_ms: min(15),
  },
  'decision_trace.day_type_decision': {
    fresh_max_ms: hr(16),
    warn_max_ms: hr(30),
    stale_max_ms: hr(72),
  },
  'decision_trace.precool_decision': {
    fresh_max_ms: hr(26),
    warn_max_ms: hr(40),
    stale_max_ms: hr(72),
  },
  'hvac.arm_mode': {
    fresh_max_ms: min(6),
    warn_max_ms: min(10),
    stale_max_ms: min(15),
  },
  'hvac.thermostat': {
    fresh_max_ms: min(12),
    warn_max_ms: min(20),
    stale_max_ms: min(30),
  },
  // 7-min fresh = controller's downgrade-recency cutoff. Cockpit mirrors
  // the scheduler so operator sees exactly the same actionability the
  // controller does. Bucket-age sawtooth typically spans 6-11 min between
  // publishes, so the freshness indicator naturally cycles green→warn→green
  // every 5-min publish cycle. Warn does NOT indicate a feed problem —
  // it indicates the controller would refuse a downgrade decision if
  // asked this tick. See spec §3.1.
  // Hand-paired with deploy/energy-stack/hvac_scheduler/freshness.py.
  'comed.prices': {
    fresh_max_ms: min(7),
    warn_max_ms: min(16),
    stale_max_ms: min(30),
  },
  'nws.forecast': {
    fresh_max_ms: min(35),
    warn_max_ms: min(90),
    stale_max_ms: hr(12),
  },
  'pjm.load_forecast': {
    fresh_max_ms: hr(14),
    warn_max_ms: hr(28),
    stale_max_ms: hr(50),
  },
  'pjm.rt_hrl_lmps': {
    fresh_max_ms: min(75),
    warn_max_ms: hr(3),
    stale_max_ms: hr(12),
  },
  // Refoss EM16P + EAGLE-3 both poll at 30s cadence. Without these
  // entries, classifyFreshness returns 'fresh' for any age, masking real
  // outages once Phase 3 wires live data.
  'refoss.channel': {
    fresh_max_ms: min(1),
    warn_max_ms: min(3),
    stale_max_ms: min(10),
  },
  'eagle.meter': {
    fresh_max_ms: min(1),
    warn_max_ms: min(3),
    stale_max_ms: min(10),
  },
}

export function classifyFreshness(source: string, ageMs: number): Freshness {
  const t = FRESHNESS_THRESHOLDS[source]
  if (!t) return 'fresh'
  if (ageMs <= t.fresh_max_ms) return 'fresh'
  if (ageMs <= t.warn_max_ms) return 'warn'
  if (ageMs <= t.stale_max_ms) return 'stale'
  return 'missing'
}
